automation/publish_signals.go

```diff
@@ -0,0 +1,258 @@
+			maps.Clear(items)
+			return nil
+		}
+	} else {
+		flush = func(integrationID string) error {
+			opts.Printf("Publish %d items...\n", len(items))
+			if opts.Verbose {
+				opts.Printf("itemsBySignal:\n")
+				opts.EncodeJSON(items)
+			}
+			result, err := c.PublishSignals(integrationID, items).Do(ctx)
+			if err != nil {
+				return err
+			}
+			if opts.Verbose {
+				opts.EncodeJSON(result)
+			}
+			total += len(items)
+			maps.Clear(items)
+			return nil
+		}
+	}
+
+	for _, id := range p.Integrations {
+		q.Skip = 0
+		more := true
+		for more {
+			if err := ctx.Err(); err != nil {
+				return err
+			}
+
+			var err error
+			more, err = p.addItems(ctx, items, c, id, q, opts)
+			if err != nil {
+				return err
+			}
+			q.Skip += q.Limit
+
+			if len(items) >= publishSignalsPageSize {
+				if err := flush(id); err != nil {
+					return err
+				}
+			}
+		}
+
+		if err := flush(id); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p PublishSignals) addItems(ctx context.Context, dest map[string]views.ItemSave, c *clarify.Client, integrationID string, q query.Query, opts PublishOptions) (bool, error) {
+	results, err := c.SelectSignals(integrationID).
+		Query(q).
+		Include("item").Do(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	// This logic only updates items if the signals has changed since last time
+	// the item was published, or if our transform version has changed.
+	prevItemsBySignal := make(map[string]views.Item, len(results.Included.Items))
+	for _, item := range results.Included.Items {
+		prevItemsBySignal[item.Meta.Annotations.Get(AnnotationPublisherSignalID)] = item
+	}
+
+	for _, signal := range results.Data {
+		// Only update item if either the source signal or transform version
+		// have changed.
+		prevItem, ok := prevItemsBySignal[signal.ID]
+		ok = ok && prevItem.Meta.Annotations.Get(AnnotationPublisherTransformVersion) == p.TransformVersion
+		ok = ok && prevItem.Meta.Annotations.Get(AnnotationPublisherSignalAttributes) == signal.Meta.AttributesHash.String()
+		if ok {
+			if opts.Verbose {
+				opts.Printf("Skip signal %s: item is up-to-date\n", signal.ID)
+			}
+			continue
+		}
+
+		item := views.PublishedItem(signal)
+		// Before running configured transforms, set the initial visible value
+		// to the value from the existing item, or true for new items.
+		if prevItem.ID != "" {
+			item.Visible = prevItem.Attributes.Visible
+		} else {
+			item.Visible = true
+		}
+
+		for _, f := range p.Transforms {
+			f(&item)
+		}
+
+		// After running configured transformations, set automation package
+		// annotations.
+		item.Annotations.Set(AnnotationPublisherName, opts.Publisher)
+		item.Annotations.Set(AnnotationPublisherTransformVersion, p.TransformVersion)
+		item.Annotations.Set(AnnotationPublisherSignalAttributes, signal.Meta.AttributesHash.String())
+		item.Annotations.Set(AnnotationPublisherSignalID, signal.ID)
+
+		dest[signal.ID] = item
+	}
+
+	more := len(results.Data) == q.Limit
```

Comment:
can you pls clarify what `more` flag actually checks?

Reply:
What we want is to fetch _all_ signals per integration, but the API only allow us to fetch 1000; so we must iterate through signals until we found all of them. This implementation just keeps iterating until there are less than 1000 items returned. This is an _alternative_ to fetching total (which we also _could_ do), that has the following trade offs:

1. We save the cost of having to do a total count per query.
2. If total % 1000 == 0, we do one additional query.

This is basically a _bet_ that `total % 1000 == 0` is uncommon. It's also a solution that would still work, if the total count ended up wrong (e.g. due to bugs our end).

I can add a comment explaining it.

Reply:
`more := reults.meta.Total > q.Skip + len(resuls.Data)`
Can also ins some cases, _not work_. E.g. if you add `skip: 1000000`, it would keep iterating for ever. So would have to be:
```
more := len(results.Data) == q.Limit &&  reults.meta.Total > q.Skip + len(resuls.Data)
```